[BE] 숙소 예약 기능 구현 #75

wooody92 opened the pull request:
### 숙소 예약 기능

* 숙소 예약 기능
  - 예약 가능 여부 체크 및 예외처리 진행
  - 예약 가능하면 해당 기간 전부 예약 진행

* 예약 취소 기능은 구현 후 업데이트 하겠습니다.

wooody92 committed:
[#75] refactor : Entity 관련 annotation 추가

- Booking.java : FetchType 및 NotNull 추가

wooody92 committed:
[#75] feat : booking add 메서드 추가

- Room.java :
  - bookings cascade type 추가
  - addBooking 메서드 추가
  - 리스트 타입 선언 시 초기화

wooody92 committed:
[#75] feat : booking constructor 및 create 메서드 추가

- Booking.java : 생성자 및 생성 메서드 추가했습니다.

wooody92 committed:
[#75] feat : exception 및 exceptionHandler 추가

- exception.package : 요청 기간에 예약이 불가하면 발생하는 예외 및 핸들러 추가했습니다.

wooody92 committed:
[#75] feat : 여러 booking을 add하는 메서드 추가

- Room.java : 기간 내 모든 날짜를 add 하는 addBookings 추가 했습니다.

wooody92 committed:
[#75] fix : booking table initial 되지 않는 버그 수정

- CREATE TABLE <IF NOT EXISTS> booking 추가했습니다.

wooody92 committed:
[#75] feat : 숙소 검색 시 유효하지 않은 아이디로 검색 시 발생하는 예외 생성

- NotFoundDataException.java

wooody92 committed:
[#75] feat : 숙소 에약 기능 추가

- RoomService.java : 숙소 예약하는 기능 추가했습니다.

sigridjineth commented on BE/src/main/java/dev/codesquad/airbnb02/domain/room/entity/Booking.java:
```diff
@@ -1,33 +1,56 @@
 
 @Entity
 @Getter
+@Setter
```
setter 필요하나요?

sigridjineth commented on BE/src/main/java/dev/codesquad/airbnb02/domain/room/entity/Booking.java:
```diff
@@ -1,33 +1,56 @@
+  @NotNull
   private int guest;
 
+  public Booking() {}
```
```@NoArgsConstructor``` 를 사용해서 일관성을 유지해주세요.

wooody92 replied:
네 수정하겠습니다.

sigridjineth commented on BE/src/main/java/dev/codesquad/airbnb02/domain/room/entity/Room.java:
```diff
@@ -54,18 +54,18 @@
-  @OneToMany(mappedBy = "room")
   @JsonIgnore
-  private List<Booking> bookings;
+  @OneToMany(mappedBy = "room", cascade = CascadeType.ALL)
```
```CascadeType.ALL``` 외의 다른 방안은 없나요?
제 생각에는 ```CascadeType.DELETE``` 정도만 있으면 될 것 같은데 어떻게 생각하세요?
아마 User 입장에서 ```CascadeType.PERSIST```을 주면 될 것이라 생각합니다.
결국 방의 정보를 업데이트 하는 것은 사용자고, Room 객체에서 변경이 일어나면 아직 알지 못하는 버그가 우려됩니다.

* 사실 PERSIST를 무서워합니다.
	- https://github.com/beadss/jpa-study/issues/24
	- https://joont92.github.io/jpa/CascadeType-PERSIST%EB%A5%BC-%ED%95%A8%EB%B6%80%EB%A1%9C-%EC%82%AC%EC%9A%A9%ED%95%98%EB%A9%B4-%EC%95%88%EB%90%98%EB%8A%94-%EC%9D%B4%EC%9C%A0/

wooody92 replied:
우선 Booking이라는 테이블이 Room과 User 중 어느 Aggregation에 속하는지부터 생각을 해봐야 할 것 같아요.
우선, 현재 상태에서는 User 테이블이 따로 존재하지않아 위와같이 처리헀는데요, Agrregation이 User라면 따로 리뷰 주신 것처럼 따로 처리를 해줘야겠네요. 저는 Room 쪽에 더 가깝다고 느껴지는데 어떻게 생각하시는지 의논했으면 합니다.

sigridjineth commented on BE/src/main/java/dev/codesquad/airbnb02/domain/room/entity/Room.java:
```diff
@@ -99,6 +99,18 @@ public boolean isValidDate(LocalDate checkin, LocalDate checkout) {
 
+  public void addBooking(Booking booking) {
+    bookings.add(booking);
+    booking.setRoom(this);
```
아 이것 때문에 Setter를 사용하셨군요.. 세터를 사용하지 않고 구현할 수 있는 방법이 있을까요?
뭔가 Room에서 booking의 정보를 세팅하는 것이 다소 이상하게 느껴져요.
결국에는 Booking도 엔티티로 처리하려고 하니까 나타나는 현상일텐데요.
엔티티로 처리한다고 해도, 서로 다른 애그리거트의 정보(Booking의 정보)를 Room에서 변경하려고 하는 것이 위험하다고 느껴집니다.

wooody92 replied:
좋은 지적입니다. 객체단위에서는 굳이 room 전체 정보를 갖고있을 필요 없이 roomId 만 갖고있도록 하는 것이 의미에 더욱 맞다고 생각이 되네요.  `@ElementCollection` 과 `@CollectionTable` 학습 후 리팩토링 해보겠습니다.

sigridjineth commented on BE/src/main/java/dev/codesquad/airbnb02/domain/room/business/RoomService.java:
```diff
@@ -37,4 +39,14 @@ public RoomService(RoomRepository roomRepository) {
+  public void createBooking(Long roomId, LocalDate checkin, LocalDate checkout) {
+    Room room = roomRepository.findById(roomId).orElseThrow(() -> new NotFoundDataException("해당 숙소를 찾을 수 없습니다."));
+    if (!room.isValidDate(checkin, checkout)) {
+      throw new BookingNotAllowedException("이미 예약된 날짜가 존재 합니다.");
```
```RuntimeException```을 상속한 별도의 커스텀 에러 생성하고 이를 처리 👍 
저는 귀찮아서 안 했어요.......

sigridjineth reviewed (changes requested):
수고 많으셨습니다. 전반적으로 깔끔한 코드라고 생각합니다. 👍 
하지만 Room 객체 내부에서 Booking의 상태를 setter로 임의 지정하는 것에 대해서는 동의하지 않습니다.
제 생각에는 Room에서 Room이 갖는 Booking 정보를 업데이트 하면서, Booking의 상태 자체도 업데이트를 해야 하니까 그렇게 설계하신 것 같은데요. 같이 좋은 방법을 논의해봤으면 합니다.

sigridjineth commented:
@wooody92 PR을 draft로 바꾸려고 하는데 어떠신가요?
예약 취소 기능이 완료되면 그 때 코드를 병합해도 좋을 것 같습니다.
위 PR에서 계속 토론해봅시다. 구현이 완료되시면 draft에서 open으로 다시 바꿔주세요. :D

wooody92 reviewed (commented):
리뷰 감사합니다. 코멘트 남겼습니다. 의논사항 같이 논의해봐요!

wooody92 committed:
[#75] feat : 예약 취소 url path 추가

- BookingController.java : 예약하기, 예약취소 url path 추가 했습니다.
- RoomController.java : BookingController로 기능 분리했습니다.

wooody92 committed:
[#75] feat : Booking 정보를 반환하는 DTO 생성

- BookingResponseDto.java : 예약 생성 및 취소시 필요한 데이터를 반환해주기 위한 dto 생성했습니다.

wooody92 committed:
[#75] feat : isEqualsBookDate 메서드 추가

- Booking.java : 예약날짜 일치하는지 조회하는 메서드 추가했습니다. JsonIgnore 및 toString exclude Booking 클래스로 변경했습니다.

wooody92 committed:
[#75] feat : 예약 취소 기능 추가

- findBookingByDate : Booking list에 해당하는 날짜가 있으면 반환하는 메서드 추가했습니다.
- removeBookings : 입력날짜에 예약이 존재하면 삭제하는 메서드 추가했습니다.

wooody92 committed:
[#75] chore : LocalDate 타입 변경으로 테스트코드 주석처리

- 테스트코드 주석처리했습니다.

wooody92 committed:
[#75] feat : 예약 취소 기능 구현

- removeBooking : 해당 기간 예약 취소 후 dto로 반환합니다.

wooody92 committed:
[#75] refactor : 반환값 void로 변경

- return void로 변경했습니다.

wooody92 committed:
[#75] refactor : 숙소 예약 및 취소 요청 메서드 양식 변경

- POST, DELETE MAPPING으로 요청양식 변경했습니다.
